token/rust/src/token.rs

```diff
@@ -0,0 +1,260 @@
+    AccountInvalidOwner,
+    #[error("invalid account mint")]
+    AccountInvalidMint,
+}
+
+pub type TokenResult<T> = Result<T, TokenError>;
+
+pub struct Token<ST, TS> {
+    client: Arc<dyn TokenClient<ST>>,
+    pubkey: Pubkey,
+    payer: TS,
+}
+
+impl<ST, TS> fmt::Debug for Token<ST, TS>
+where
+    TS: Signer,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Token")
+            .field("pubkey", &self.pubkey)
+            .field("payer", &self.payer.pubkey())
+            .finish()
+    }
+}
+
+impl<ST, TS> Token<ST, TS>
+where
+    ST: SendTransaction,
+    TS: Signer,
+{
+    async fn process_ixs<T: Signers>(
+        client: &Arc<dyn TokenClient<ST>>,
+        payer: &TS,
+        instructions: &[Instruction],
+        signing_keypairs: &T,
+    ) -> TokenResult<ST::Output> {
+        let recent_blockhash = client
+            .get_recent_blockhash()
+            .await
+            .map_err(TokenError::Client)?;
+
+        let mut tx = Transaction::new_with_payer(instructions, Some(&payer.pubkey()));
+        tx.try_partial_sign(&[payer], recent_blockhash)
+            .map_err(|error| TokenError::Client(error.into()))?;
+        tx.try_sign(signing_keypairs, recent_blockhash)
+            .map_err(|error| TokenError::Client(error.into()))?;
+
+        client
+            .send_transaction(tx)
+            .await
+            .map_err(TokenError::Client)
+    }
+
+    pub fn new(client: Arc<dyn TokenClient<ST>>, address: Pubkey, payer: TS) -> Self {
+        Token {
+            client,
+            pubkey: address,
+            payer,
+        }
+    }
+
+    /// Get token address.
+    pub fn get_address(&self) -> &Pubkey {
+        &self.pubkey
+    }
+
+    /// Create and initialize a token.
+    pub async fn create_mint<'a, S: Signer>(
+        client: Arc<dyn TokenClient<ST>>,
+        payer: TS,
+        mint_account: &'a S,
+        mint_authority: &'a Pubkey,
+        freeze_authority: Option<&'a Pubkey>,
+        decimals: u8,
+    ) -> TokenResult<Self> {
+        Self::process_ixs(
+            &client,
+            &payer,
+            &[
+                system_instruction::create_account(
+                    &payer.pubkey(),
+                    &mint_account.pubkey(),
+                    client
+                        .get_minimum_balance_for_rent_exemption(state::Mint::LEN)
+                        .await
+                        .map_err(TokenError::Client)?,
+                    state::Mint::LEN as u64,
+                    &spl_token::id(),
+                ),
+                instruction::initialize_mint(
+                    &spl_token::id(),
+                    &mint_account.pubkey(),
+                    mint_authority,
+                    freeze_authority,
+                    decimals,
+                )?,
+            ],
+            &[mint_account],
+        )
+        .await?;
+
+        Ok(Self::new(client, mint_account.pubkey(), payer))
+    }
+
```

Comment:
Does it make sense to include a `payer: Option<S>` input to the methods below to allow overriding the Token default fee-payer?

Reply:
I added functions looking on the JS API and there is no optional payer. We can add function:
```rust
fn with_payer<S2: Signer>(&self, payer: S2) -> Token<T, S2> {
  Token { client: Arc::clone(&self.client), pubkey: self.pubkey, payer }
}
```
so it will be possible to do something like `token.with_payer(new_payer_here).create_associated_token_account(...)`.

Reply:
I don't think we need to be beholden to the JS api, do we?
But anyway `with_payer()` should work fine. Thanks!